[13791] Adding content filter documentation (#361)
eProsima/Fast-DDS-docs: docs/fastdds/dds_layer/topic/contentFilteredTopic/defaultFilter.rst

This section shows this default SQL-like syntax and how to use it.

* :ref:`default_sql_filter_grammar`
* :ref:`default_sql_filter_like`
* :ref:`default_sql_filter_match`
* :ref:`default_sql_filter_type_comparisons`
* :ref:`default_sql_filter_example`

.. _default_sql_filter_grammar:

Grammar
-------

The allowed SQL expressions are defined with the BNF-grammar below.

The following conventions are made:

- "Terminals" are quoted.
- ``TOKENS`` are typeset in code block and small caps.

.. productionlist::
    Expression: FilterExpression
    FilterExpression: Condition
    Condition: Predicate | Condition "AND" Condition | Condition "OR" Condition | "NOT" Condition | "(" Condition ")"
    Predicate: ComparisonPredicate | BetweenPredicate
    ComparisonPredicate: FIELDNAME RelOp Parameter | Parameter RelOp FIELDNAME | FIELDNAME RelOp FIELDNAME
    BetweenPredicate: FIELDNAME "BETWEEN" Range | FIELDNAME "NOT BETWEEN" Range
    RelOp: "=" | ">" | ">=" | "<" | "<=" | "<>" | "!=" | `like` | `match`
    Range: Parameter "AND" Parameter
    Parameter: INTEGERVALUE | CHARVALUE | FLOATVALUE | STRINGVALUE | ENUMERATEDVALUE | PARAMETER

> 

Reviewer: Remove extra blank line

Reviewer 2: done on 2eaf78ee

> 

Reviewer: We should include a note stating that both the capitalized and lowercase special words are supported: `AND`, `and`, `OR`, `or`, `LIKE`, etc.

Reviewer 2: done on 2eaf78ee

The syntax and meaning of the tokens used in the SQL grammar is described as follows:

- **FIELDNAME**: is a reference to a field in the data-structure.
  The dot ``.`` is used to navigate through nested structures.
  The number of dots that may be used in a FIELDNAME is unlimited.
  The FIELDNAME can refer to fields at any depth in the data structure.
  The names of the field are those specified in the IDL definition of the corresponding structure.

  .. productionlist::
      FIELDNAME: FieldNamePart ( "." FieldNamePart )*
      FieldNamePart: Identifier ( "[" Integer "]" )?

  An example of FIELDNAME:

  .. tabs::

      .. tab:: Filter expression

          .. code::

              "points[0] = 0 AND color.red < 100"

      .. tab:: Associated IDL

          .. code:: IDL

              struct Color
              {
                  octet red;
                  octet green;
                  octet blue;
              };

              struct Shape
              {
                  long points[4];
                  Color color;
              };

> 

Reviewer: Remove extra blank line

Reviewer 2: done on 2eaf78ee

- **INTEGERVALUE**: Any series of digits, optionally preceded by a plus or minus sign, representing a decimal integer
  value within the range of the system. A hexadecimal number is preceded by 0x and must be a valid hexadecimal
  expression.

  .. productionlist::
      INTEGERVALUE: (["+","-"])? Integer
      Integer: (["0"-"9"])+ | ["0x","0X"](["0"-"9", "A"-"F", "a"-"f"])+

  An example of INTEGERVALUE:

  .. code::

      value = -10

- **CHARVALUE**: A single character enclosed between single quotes.

  .. productionlist::
      CHARVALUE: "'" Character "'"
      Character: ~["\n"]

  An example of CHARVALUE:

  .. code::

      value = 'c'

- **FLOATVALUE**: Any series of digits, optionally preceded by a plus or minus sign and optionally including a floating

>   point (‘.’).

Reviewer: ```suggestion
  point (``.``).
```

  Enumerated values consist of the name of the enumeration label enclosed in single quotes.
  The name used for the enumeration label must correspond to the label names specified in the IDL definition of the
  enumeration.

  .. productionlist::
      ENUMERATEDVALUE: ["'"] ~["'", "\r", "\n"] ["'"]

  An example of ENUMERATEDVALUE:

  .. tabs::

      .. tab:: Filter expression

          .. code::

              value = 'ENUM_VALUE_1'

      .. tab:: Associated IDL

          .. code::

              enum MyEnum
              {
                  ENUM_VALUE_1,
                  ENUM_VALUE_2,
                  ENUM_VALUE_3
              };

              struct Enumerators
              {
                  MyEnum value;
              };

- **PARAMETER**: A parameter is of the form ``%n``, where ``n`` represents a natural number (zero included) smaller than
  100.
  It refers to the ``n + 1 th`` argument in the given context.

  .. productionlist::
      PARAMETER: ["%"] ["0"-"9"] (["0"-"9"])?

  An example of PARAMETER:

  .. code::

      value = %1

.. _default_sql_filter_like:

Like condition
--------------

> The ``like`` operator is similar as the one defined by SQL.

Reviewer 2: done in e50457f3c6f29b

This operator only can be used with strings.
There are two wildcards could be used in conjunction with this operator

- The percent sign ``%`` (or its alias ``*``) represents zero, one, or multiple characters.
- The underscore sign ``_`` (or its alias ``?``) represents one single character.

All wildcards can also be used in combinations.

An example of ``like`` operator

.. tabs::

    .. tab:: Filter expression

        .. code::

            "str like '%bird%'"

    .. tab:: Associated IDL

        .. code:: IDL

            struct Like
            {
                string str;
            };

> where string "There are birds flying" will be return ``true``.

Reviewer: ```suggestion
where string ``There are birds flying`` will return ``true``.
```

Reviewer: This suggestion is not completely applied.

```suggestion
where string ``There are birds flying`` will return ``true``.
```

The verb `be` is not needed.

.. _default_sql_filter_match:

Match condition
---------------

The ``match`` operator performs a full-text search using a regular expression.

> It uses the Basic Regular Expression (BRE) defined by POSIX.
> More information can be found
> `here <https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_03>`_

Reviewer: ```suggestion
It uses the `Basic Regular Expression (BRE) defined by POSIX <https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_03>`_.
```

An example of ``match`` operator

.. tabs::

    .. tab:: Filter expression

        .. code::

            "str match '^The'"

    .. tab:: Associated IDL

        .. code:: IDL

            struct Like
            {
                string str;
            };

> where string "There are birds flying" will be return ``true``.

Reviewer: ```suggestion
where string ``There are birds flying`` will return ``true``.
```

Reviewer: The verb `be` should be removed.

.. _default_sql_filter_type_comparisons:

Type comparisons
-----------------

For the supported operators in the grammar, next table shows the type compatibility.

.. list-table::
    :header-rows: 1
    :stub-columns: 1

    * - Operator1 | Operator2
      - BOOLEAN
      - INTEGER
      - FLOAT
      - CHAR
      - STRING
      - ENUM
    * - BOOLEAN
      - ✅
      - ✅
      - ❌
      - ❌
      - ❌
      - ❌
    * - INTEGER
      - ✅
      - ✅
      - ✅
      - ❌
      - ❌
      - ❌
    * - FLOAT
      - ❌
      - ✅
      - ✅
      - ❌
      - ❌
      - ❌
    * - CHAR
      - ❌
      - ❌
      - ❌
      - ✅
      - ✅
      - ❌
    * - STRING
      - ❌
      - ❌
      - ❌
      - ✅
      - ✅
      - ❌
    * - ENUM
      - ❌
      - ✅
      - ❌
      - ❌
      - ❌
      - ✅ *

**(*)** Only for the same enumerated type.

.. _default_sql_filter_example:

Example
-------

> Assuming Topic "Shape" has next IDL definition.

Reviewer: ```suggestion
Assuming Topic ``Shape`` has next IDL definition:
```

Reviewer 2: done on 2eaf78ee

.. code:: IDL

   struct Shape
   {
       long x,
       long y,
       long z,
       long width,
       long height
   };

An example of filter expression would be:

.. code::

    "x < 23 AND y > 50 AND width BETWEEN %0 AND %1"

A :ref:`dds_layer_topic_contentFilteredTopic` may be created using this filter expression as explained in section
:ref:`dds_layer_topic_contentFilteredTopic_creation`.

.. literalinclude:: /../code/DDSCodeTester.cpp
   :language: c++
   :start-after: //DDS_CONTENT_FILTERED_TOPIC_SQL_EXAMPLE
   :end-before: //!
   :dedent: 8

In this example parameters are used.

> Internally the :ref:`dds_layer_topic_contentFilteredTopic` will be created with next filter expression, after settings

Reviewer: ```suggestion
Internally the :ref:`dds_layer_topic_contentFilteredTopic` will be created with the filter expression below, after setting
```

Reviewer 2: done on 2eaf78ee

  A string starts with a left or right quote, but ends with a right quote.

  .. productionlist::
      STRINGVALUE: ["'"] ~["'", "\r", "\n"] ["'"]

  An example of STRINGVALUE:

  .. code::

      value = 'This is a string'

- **ENUMERATEDVALUE**: An enumerated value is a reference to a value declared within an enumeration.
  Enumerated values consist of the name of the enumeration label enclosed in single quotes.
  The name used for the enumeration label must correspond to the label names specified in the IDL definition of the
  enumeration.

  .. productionlist::
      ENUMERATEDVALUE: ["'"] ~["'", "\r", "\n"] ["'"]

  An example of ENUMERATEDVALUE:

  .. tabs::

      .. tab:: Filter expression

          .. code::

              value = 'ENUM_VALUE_1'

      .. tab:: Associated IDL

          .. code::

              enum MyEnum
              {
                  ENUM_VALUE_1,
                  ENUM_VALUE_2,
                  ENUM_VALUE_3
              };

              struct Enumerators
              {
                  MyEnum value;
              };

- **BOOLEANVALUE**: Can either be `true` of `false`, case sensitive.

  .. productionlist::

>       BOOLEANVALUE: ["TRUE", "true", "FALSE", "false"]

Reviewer: I suggest following the same order as in the grammar before. So `BOOLEANVALUE` should be before `INTEGERVALUE`.

- **PARAMETER**: A parameter is of the form ``%n``, where ``n`` represents a natural number (zero included) smaller than
  100.
  It refers to the ``n + 1 th`` argument in the given context.

  .. productionlist::
      PARAMETER: ["%"] ["0"-"9"] (["0"-"9"])?

  An example of PARAMETER:

  .. code::

      value = %1

.. _default_sql_filter_like:

Like condition
--------------

The `like <https://www.w3schools.com/sql/sql_like.asp>`_ operator is similar as the one defined by SQL.

> This operator only can be used with strings.

Reviewer: ```suggestion
This operator can only be used with strings.
```

There are two wildcards that could be used in conjunction with this operator

- The percent sign ``%`` (or its alias ``*``) represents zero, one, or multiple characters.
- The underscore sign ``_`` (or its alias ``?``) represents one single character.

All wildcards can also be used in combinations.

An example of ``like`` operator

.. tabs::

    .. tab:: Filter expression

        .. code::

            "str like '%bird%'"

    .. tab:: Associated IDL

        .. code:: IDL

            struct Like
            {
                string str;
            };

where string ``There are birds flying`` will be return ``true``.

.. _default_sql_filter_match:

Match condition
---------------

The ``match`` operator performs a full-text search using a regular expression.

> This operator only can be used with strings.

Reviewer: ```suggestion
This operator can only be used with strings.
```

.. include:: ../../../../03-exports/aliases.include
.. include:: ../../../../03-exports/aliases-api.include
.. include:: ../../../../03-exports/roles.include

.. _dds_layer_topic_contentFilteredTopic_default_filter:

The default SQL-like filter
===========================

Filter expressions used by :ref:`dds_layer_topic_contentFilteredTopic` API may use a subset of SQL syntax, extended with

> the possibility to use program variables in the SQL expression. |br|

Reviewer 2: done on 2eaf78ee

> the possibility to use program variables in the SQL expression. |br|

Reviewer: After finishing a sentence, the next sentence should start at a new line. It is easier to review and make suggestions.

This section shows this default SQL-like syntax and how to use it.

* :ref:`default_sql_filter_grammar`
* :ref:`default_sql_filter_like`
* :ref:`default_sql_filter_match`
* :ref:`default_sql_filter_type_comparisons`
* :ref:`default_sql_filter_example`

.. _default_sql_filter_grammar:

Grammar
-------

The allowed SQL expressions are defined with the BNF-grammar below.

The following conventions are made:

- "Terminals" are quoted.

> - ``TOKENS`` are typeset in code block and small caps.

Reviewer 2: done on 2eaf78ee

The syntax and meaning of the tokens used in the SQL grammar is described as follows:

- **FIELDNAME**: is a reference to a field in the data-structure.
  The dot ``.`` is used to navigate through nested structures.
  The number of dots that may be used in a FIELDNAME is unlimited.
  The FIELDNAME can refer to fields at any depth in the data structure.
  The names of the field are those specified in the IDL definition of the corresponding structure.

  .. productionlist::
      FIELDNAME: FieldNamePart ( "." FieldNamePart )*
      FieldNamePart: Identifier ( "[" Integer "]" )?

  An example of FIELDNAME:

  .. tabs::

      .. tab:: Filter expression

          .. code::

              "points[0] = 0 AND color.red < 100"

      .. tab:: Associated IDL

          .. code:: IDL

              struct Color
              {
                  octet red;
                  octet green;
                  octet blue;
              };

              struct Shape
              {
                  long points[4];
                  Color color;
              };

- **INTEGERVALUE**: Any series of digits, optionally preceded by a plus or minus sign, representing a decimal integer

>   value within the range of the system. A hexadecimal number is preceded by 0x and must be a valid hexadecimal

Reviewer 2: Yes, a it was already merged.

>   expression.

Reviewer 2: done in e50457f3c6f29b

  100.
  It refers to the ``n + 1 th`` argument in the given context.

  .. productionlist::
      PARAMETER: ["%"] ["0"-"9"] (["0"-"9"])?

  An example of PARAMETER:

  .. code::

      value = %1

.. _default_sql_filter_like:

Like condition
--------------

The ``like`` operator is similar as the one defined by SQL.
This operator only can be used with strings.
There are two wildcards could be used in conjunction with this operator

- The percent sign ``%`` (or its alias ``*``) represents zero, one, or multiple characters.
- The underscore sign ``_`` (or its alias ``?``) represents one single character.

All wildcards can also be used in combinations.

An example of ``like`` operator

.. tabs::

    .. tab:: Filter expression

        .. code::

            "str like '%bird%'"

    .. tab:: Associated IDL

        .. code:: IDL

            struct Like
            {
                string str;
            };

where string "There are birds flying" will be return ``true``.

.. _default_sql_filter_match:

Match condition
---------------

The ``match`` operator performs a full-text search using a regular expression.
It uses the Basic Regular Expression (BRE) defined by POSIX.

> More information can be found

Reviewer: I think we should also indicate that this operator is intended to be used with strings.

`here <https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_03>`_

An example of ``match`` operator

.. tabs::

    .. tab:: Filter expression

        .. code::

            "str match '^The'"

    .. tab:: Associated IDL

        .. code:: IDL

            struct Like
            {
                string str;
            };

where string "There are birds flying" will be return ``true``.

.. _default_sql_filter_type_comparisons:

Type comparisons
-----------------

For the supported operators in the grammar, next table shows the type compatibility.

.. list-table::
    :header-rows: 1
    :stub-columns: 1

    * - Operator1 | Operator2

>       - BOOLEAN

Reviewer: This type is the first time is mentioned in the grammar and it has not been defined before. I do not know if a BOOLEANVALUE should be included in the grammar section.

.. productionlist::
    Expression: FilterExpression
    FilterExpression: Condition

>     Condition: Predicate | Condition "AND" Condition | Condition "OR" Condition |

Reviewer: I am not sure if this suggestion improves readability/presentation

![Grammar](https://user-images.githubusercontent.com/69244257/162204663-8c66103d-6ea3-4de5-9e26-a05d115830e7.png)
